src/Rings/oscar_singular.jl

```diff
@@ -0,0 +1,458 @@
+Singular ring `S[x_1,...,x_n]` with `R` isomorphic to `S`.
+"""
+iso_oscar_singular_poly_ring
+
+abstract type OscarSingularCoefficientRingMap{S, T} <: Map{S, T, Any, Any} end
+
+domain(f::OscarSingularCoefficientRingMap) = f.R
+
+codomain(f::OscarSingularCoefficientRingMap) = f.S
+
+# fallback
+function image(f::OscarSingularCoefficientRingMap, x)
+  parent(x) !== domain(f) && error("Element not in domain")
+  return codomain(f)(x)
+end
+
+(f::OscarSingularCoefficientRingMap)(x) = image(f, x)
+
+function preimage(f::OscarSingularCoefficientRingMap, x)
+  parent(x) !== codomain(f) && error("Element not in codomain")
+  return domain(f)(x)
+end
+
+# ZZ
+struct OscarSingularCoefficientRingMapZZ <: OscarSingularCoefficientRingMap{ZZRing, Singular.Integers} 
+  R::ZZRing # singleton, but keep for consistency
+  S::Singular.Integers
+end
+
+iso_oscar_singular_coeff_ring(R::ZZRing) = OscarSingularCoefficientRingMapZZ(R, Singular.Integers())
+
+# conversion is unambigous via parent object call
+
+# QQ
+struct OscarSingularCoefficientRingMapQQ <: OscarSingularCoefficientRingMap{QQField, Singular.Rationals}
+  R::QQField # singleton, but keep for consistency
+  S::Singular.Rationals
+end
+
+iso_oscar_singular_coeff_ring(R::QQField) = OscarSingularCoefficientRingMapQQ(R, Singular.Rationals())
+
+# prime field, small characteristic
+struct OscarSingularCoefficientRingMapfpField <: OscarSingularCoefficientRingMap{fpField, Singular.N_ZpField}
+  R::fpField
+  S::Singular.N_ZpField
+end
+
+iso_oscar_singular_coeff_ring(R::fpField) = OscarSingularCoefficientRingMapfpField(R, Singular.Fp(Int(characteristic(R))))
+
+# ZZ/nZZ, n small and big
+struct OscarSingularCoefficientRingMapZZMod{U} <: OscarSingularCoefficientRingMap{U, Singular.N_ZpField}
+  R::U
+  S::Singular.N_ZnRing
+end
+
+iso_oscar_singular_coeff_ring(R::U) where {U <: Union{zzModRing, ZZModRing}} = OscarSingularCoefficientRingMapZZMod{U}(R, Singular.residue_ring(Singular.Integers(), BigInt(modulus(R)))[1]
+)
+
+# QQ(a)
+struct OscarSingularCoefficientRingMapAbsSimpleNumField <: OscarSingularCoefficientRingMap{AbsSimpleNumField, Singular.N_ZpField}
+  R::AbsSimpleNumField
+  S::Singular.N_AlgExtField
+end
+
+function iso_oscar_singular_coeff_ring(R::AbsSimpleNumField)
+  minpoly = defining_polynomial(R)
+  Qa = parent(minpoly)
+  a = gen(Qa)
+  SQa, (Sa,) = Singular.FunctionField(Singular.QQ, _variables_for_singular(symbols(Qa)))
+  Sminpoly = SQa(coeff(minpoly, 0))
+  for i in 1:degree(minpoly)
+    Sminpoly += SQa(coeff(minpoly, i))*Sa^i
+  end
+  SK, _ = Singular.AlgebraicExtensionField(SQa, Sminpoly)
+  return OscarSingularCoefficientRingMapAbsSimpleNumField(R, SK)
+end
+
+# It is nonsense, but the conversion resides in src/number/n_algExt.jl
+# via parent object call overloading
+
+# GF(p, n), small p
+
+struct OscarSingularCoefficientRingMapfqPolyRepField <: OscarSingularCoefficientRingMap{fqPolyRepField, Singular.N_ZpField}
+  R::fqPolyRepField
+  S::Singular.N_AlgExtField
```

Comment:
OK now I am really lost, and the comment confuses me further.... So this is a coefficient map where the codomain is declared in the type signature as having type `N_ZpField` but `S` (the actual codomain?) has type `N_AlgExtField`. But there is  comment which says this "is nonsense" so perhaps this is intending to say it is on purpose? Or what exactly is nonsense?

Reply:
The nonsense is that `R(S(1))` is working. The implementation is not very safe and silently gives wrong results for wrong input. The thing with the type mismatch is just a copy & paste error.